clang/lib/CIR/CodeGen/CIRGenFunction.cpp

```diff
@@ -242,12 +242,15 @@ void CIRGenFunction::LexicalScope::cleanup() {
     }
   };
 
-  if (returnBlock != nullptr) {
-    // Write out the return block, which loads the value from `__retval` and
-    // issues the `cir.return`.
+  // Cleanup are done right before codegen resumes a scope. This is where
+  // objects are destroyed. Process all return blocks.
```

Comment:
In spite of what the comment says, I think this still has problems if we are returning through a cleanup. In the upstream code, I haven't yet committed the patch that handles return through cleanups (https://github.com/llvm/llvm-project/pull/163849), but I don't think that will work with multiple cleanups, as this case fails in the incubator:

https://godbolt.org/z/dTjjcG38d

That's not necessarily a problem with this PR, just a bigger problem that we need to solve with regard to cleanups.

Reply:
+1